### PR TITLE
Allow to disable http2 for GCS.

### PR DESCRIPTION
**Pull request by cyriltovena**
This also tweaks our idle connection pooling across the board and reduce the insane parallelism when fetching chunks.

This is the result of an experiment I run in our cluster.

Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (dannykopping) [inline pkg/storage/chunk/azure/blob_storage_client.go:71]: Should these not all be configurable in S3/Azure Blob/GCS/etc? Seems like something we can move to the common storage config.

These changes might not be appropriate for everyone, and this is could be a fairly substantial change for some folks in some environments - not having a knob to turn would be unfortunate.

File: pkg/storage/chunk/gcp/gcs_object_client.go
Comment (dannykopping) [inline pkg/storage/chunk/gcp/gcs_object_client.go:52]:
```suggestion
	f.BoolVar(&cfg.EnableOpenCensus, prefix+"gcs.enable-opencensus", true, "Enable OpenCensus (OC) instrumentation for all requests.")
	f.BoolVar(&cfg.EnableHTTP2, prefix+"gcs.enable-http2", true, "Enable HTTP2 connections.")
```

File: pkg/storage/chunk/gcp/instrumentation.go
Comment (dannykopping) [inline pkg/storage/chunk/gcp/instrumentation.go:59]: Can you please link the documentation here?

File: pkg/storage/chunk/util/parallel_chunk_fetch.go
Comment (dannykopping) [inline pkg/storage/chunk/util/parallel_chunk_fetch.go:14]: This should probably also be configurable if we're going to make such a drastic change

**Review by dannykopping — COMMENTED** [review]
Looks good; I think it'd be valuable to share some of the benchmarks we collected when testing this out - your PR description doesn't really describe the huge benefits this can have. Let's shout about it :slightly_smiling_face:

File: pkg/storage/chunk/util/parallel_chunk_fetch.go
Comment (cyriltovena) [reply pkg/storage/chunk/util/parallel_chunk_fetch.go:14]: I can't see a situation where above 150 is making sense. I'm actually stopping the bleed with this, we rarely download above 50 chunk in parallel but when we do we should not go higher than 150 per request otherwise we'll starve the process from goroutine. I'd prefer to avoid a configuration for this.

**Comment by cyriltovena** [comment]
> Looks good; I think it'd be valuable to share some of the benchmarks we collected when testing this out - your PR description doesn't really describe the huge benefits this can have. Let's shout about it 🙂

I was able to 2x my throughput but I'd prefer to run this in a real environment first to not oversell it.

File: pkg/storage/chunk/util/parallel_chunk_fetch.go
Comment (dannykopping) [reply pkg/storage/chunk/util/parallel_chunk_fetch.go:14]: To be clear: I'm not suggesting this value is incorrect or not optimal, but since we may want to tune this in the future it should at least be an (undocumented, perhaps) tuneable via config IMHO. Won't block the merge for this but if we're changing it now we may want to change it again.

Comment (cyriltovena) [reply pkg/storage/chunk/util/parallel_chunk_fetch.go:14]: Sounds good I'll see how easy it is.

File: CHANGELOG.md
Comment (KMiller-Grafana) [inline CHANGELOG.md:23]:
```suggestion
* [4942](https://github.com/grafana/loki/pull/4942) **cyriltovena**: Allow to disable HTTP/2 for GCS.
```

File: docs/sources/configuration/_index.md
Comment (KMiller-Grafana) [inline docs/sources/configuration/_index.md:734]:
```suggestion
# Enable HTTP/2 when connecting to GCS.
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (cyriltovena) [reply pkg/storage/chunk/azure/blob_storage_client.go:71]: After more thought I doubt this needs to be a knob, I'll reduce the value to 200 which seems to be what most OSS project are using.

If in the future someone needs to tune this, I'll happily add the config, but I'm not sure the effort is required right now.

File: pkg/storage/chunk/util/parallel_chunk_fetch.go
Comment (cyriltovena) [reply pkg/storage/chunk/util/parallel_chunk_fetch.go]: Done

**Review by dannykopping — APPROVED** [review]
LGTM :+1:

**Comment by cyriltovena** [comment]
@KMiller-Grafana I'll need approval for merging, Thanks !

**Review by KMiller-Grafana — APPROVED** [review]
Thanks for the changes.  LGTM.

**Pull request merged**
